// oneramp/dist/src/utils/address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const addresses = {
    celo: {
        contract: "0x1fC72f58a675ac93980E597cE6Da531d40b24c60",
        usdt: "0x02De4766C272abc10Bc88c220D214A26960a7e92",
        stable: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
        dai: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
    },
    alfajores: {
        contract: "0x1fC72f58a675ac93980E597cE6Da531d40b24c60",
        usdt: "0x02De4766C272abc10Bc88c220D214A26960a7e92",
        stable: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
        dai: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
    },
    bsc: {
        contract: "0x0CcB0071e8B8B716A2a5998aB4d97b83790873Fe",
        usdt: "0x02De4766C272abc10Bc88c220D214A26960a7e92",
        stable: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
        dai: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
    },
    bscTestnet: {
        contract: "0x0CcB0071e8B8B716A2a5998aB4d97b83790873Fe",
        usdt: "0x02De4766C272abc10Bc88c220D214A26960a7e92",
        stable: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
        dai: "0xc0EBB770F2c9CA7eD0dDeBa58Af101695Cf1BDc1",
    }
};
exports.default = addresses;
